### DISCUSSION TITLE
How to stop recording at a certain time?

### DISCUSSION THREAD

**Discussion by amajed172** [gopro/labs#1058] [Q&A] [answered]
Hello,
GoPro HERO11 Black - v2.30.70, March 2024
No battery, hooked directly to a Powerbank via USB-C

I am using Time-lapse Video to record at anytime I start it. And I want it to stop recording at 19:00 everyday. I sometimes start recording at 3:00 and sometimes at 7:00 but I want it to always stop recording at 19:00. (repeat that daily).

How do I do that?
Please help me out. Thanks.

My issue right now is that the Powerbank depleted and the camera turns off while recording, and the file gets corrupted.
The Powerbank can stay ON for about 20 Hours. I always replace it with a charged one around 3:00 and hit recording. It can easily stay ON till 19:00 (I don't need to record anything after that).

How to pull my current script and share it here so you can have a look?

Thanks for reading this.
-Anas

**Comment by dnewman-gpsw** [comment]
First, the file is not corrupted, just not closed. The next time the camera is powered on it will "repair" unclosed files, by closing them (writing the MP4 index.)

You could run this as a boot script:
```>r0!19:00E!N!R```

In pseudo code:

loop:
if(recording) 
   EndAt(19:00);
Pause();
goto loop;

**Reply by amajed172** [reply]
> First, the file is not corrupted, just not closed. The next time the camera is powered on it will "repair" unclosed files, by closing them (writing the MP4 index.)
> 
> You could run this as a boot script:
> 
> > r0!19:00E!N!R
> 
> In pseudo code:
> 
> loop: if(recording) EndAt(19:00); Pause(); goto loop;

Thanks
I tried this on my GoPro 12 Black and it didn't work
Latest firmware and latest firmware labs

used this script:
GoProQR: mTp.10r5XfW0td1r0!19:05E!N!R

I started recording on 19:01 hoping it will stop recording on 19:05 but it did not. I stopped recording manually on 19:07

Any ideas why it did not stop at 19:05 ?


Sorry for the late response. Did not have time to test things out till now

**Reply by dnewman-gpsw** [reply]
There is a typo use ```>r0``` not just ```r0```. See docs here https://gopro.github.io/labs/control/tech/

**Reply by amajed172** [reply]
> There is a typo use `>r0` not just `r0`. See docs here https://gopro.github.io/labs/control/tech/

Thanks, but still did not stop recording

I used the following script:
GoProQR: mTp.10r5XoW1!MTUSB=1>r0!07:16E!N!R


And it did not stop recording at 07:16

**Reply by dnewman-gpsw** [reply]
Note: never combine !Mwxyz commands inside a loop. It will work, just not recommended.

I have confirmed ```>r0!21:30E!R``` does absolute works as intended.

Try scanning a simpler version: 
 ![image](https://github.com/user-attachments/assets/93ab04a2-33f3-4bce-9499-b52e0978a0ad)

This will print the loop count.  Once you press record, the 60s End timer will automatically activate, and end after 60s, then the loop will continue.

**Reply by amajed172** [reply]
> Note: never combine !Mwxyz commands inside a loop. It will work, just not recommended.
> 
> I have confirmed `>r0!21:30E!R` does absolute works as intended.
> 
> Try scanning a simpler version: ![image](https://private-user-images.githubusercontent.com/10856084/367559427-93ab04a2-33f3-4bce-9499-b52e0978a0ad.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYzNzUwMzQsIm5iZiI6MTcyNjM3NDczNCwicGF0aCI6Ii8xMDg1NjA4NC8zNjc1NTk0MjctOTNhYjA0YTItMzNmMy00YmNlLTk0OTktYjUyZTA5NzhhMGFkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE1VDA0MzIxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU0OTVhOTJiMTlhMGJhZGQ1MDIxNmM5ODkyZDYyNTIyMDU1MzI0NGI3NTZkZDQ4ODVjOTdkOWMwMDcyMmYyNWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.1dCsg78fGSwJQozxZ9k8akerG5oSc0fM2JgdeEaCGJ8)
> 
> This will print the loop count. Once you press record, the 60s End timer will automatically activate, and end after 60s, then the loop will continue.

I have absuloty no idea what you just said, I also looked up the doc link you provided to try and understand but I guess I'm just stupid.. I can't really tell how to get a working script that does the following:

- Video Timelapse mode when turning on the camera (does not start recording, only ready to record when pressing record)
- 8x7 5K Res
- WiFi ON
- Timelapse Interval 10s
- TUSB to run Camera on Powerbank without a battery
- Stop recording at 19:00 no matter what time I start the recording
- A working script for HERO11 and HERO12 (I have both on a working site)

How do I create a script for all of that?

Sorry for my stupidness, I need help
Thanks

**Reply by dnewman-gpsw** [reply]
Did you try the above QR code? It was for learning, and showing the condition and end commands are working.  

You just need learn each step at a time.  You will not get the code correct by learning it all at once.

Also photo timelapse doesn't require resolution settings, as there are none, photo are always the sensor resolution. 

*TUSB=1 is scanned only once (same as old style !MTUSB=1.) You never need to scan it again. You had it is a loop, doing a one time operation, potentially thousands of times.  Look up Permanent metadata storage. 

Don't try do everything in one QR Code.

Just set the timelapse mode up manually so save yourself effort on learning that code. Then set the camera to boot into timelapse by default. This is on the standard non-Labs firmware.

Now the only command you need is:
```>r0<19:00!19:00E!N!R```

It works. Add with the correct camera setup, it does everything on your list (basically "Stop recording at 19:00 no matter what time I start the recording".)  

P.S. You haven't defined what happens if you start recording after 7pm.  I added code to ignore those captures.

**Reply by amajed172** [reply]
> Did you try the above QR code? It was for learning, and showing the condition and end commands are working.
> 
> You just need learn each step at a time. You will not get the code correct by learning it all at once.
> 
> Also photo timelapse doesn't require resolution settings, as there are none, photo are always the sensor resolution.
> 
> *TUSB=1 is scanned only once (same as old style !MTUSB=1.) You never need to scan it again. You had it is a loop, doing a one time operation, potentially thousands of times. Look up Permanent metadata storage.
> 
> Don't try do everything in one QR Code.
> 
> Just set the timelapse mode up manually so save yourself effort on learning that code. Then set the camera to boot into timelapse by default. This is on the standard non-Labs firmware.
> 
> Now the only command you need is: `>r0<19:00!19:00E!N!R`
> 
> It works.

Sorry I meant Video Timelapse mode.

Yes I did scan the QR code above, the camera started counting nonstop. I started recording and the counting stopped. then after a while the camera stopped recording. I really did not understand the loop thingy. You wanted to teach me something but I'm so in the dark that I do not see the point of that QR Code.

My issue is that I have 14 Cameras in a work site (10 HERO11 Cameras and 4 HERO12 Cameras) They are all mounted already on a place that it's pretty hard to reach without a tall ladder. so a single QR Code will make it a lot easier to setup. All of them will want to run on a Video Timelapse That stops at 19:00. all of them will run on a Powerbank.

So your suggestion is to setup the camera manually using the camera's menu for video timelapse settings and other stuff. then apply `>r0<19:00!19:00E!N!R` So it stops recording at 19:00

> P.S. You haven't defined what happens if you start recording after 7pm. I added code to ignore those captures.

If I start recording at 19:01 I want the camera to keep recording until 19:00 the next day. so 23:59 hours.
Basically I want to start recording whenever. And always stop recording when the clock hits 19:00


I really appreacte what you are doing for me. And I do want to learn. I read the whole doc but I could not connect the dots. I don't understand how scripts work. And how the camera read them. I don't understand when the camera not liking two commands together. Nothing makes sense.

**Reply by dnewman-gpsw** [reply, marked as answer]
If the camera where configure the only command you need is

```>r0!19:00E!N!R```  That supports start at 19:01 an recording to the next 19:00

Some on your team must have done some coding.  The scripting is like most computer languages, although the is simplify to fit in smaller QR Codes. The loop example simply added a print command, so you could set what was happening, and that the code is working. 

Show all the cameras ```*TUSB=1``` once. As a separate QR Code.
Nice make sure there clocks are set correctly, as they can drift. Use https://gopro.github.io/labs/control/precisiontime/

Then as you are unable to use the menus, this is an example of the full command.

```mTp.10fWoW1>r0!19:00E!N!R11```

This was the code does, on step at a time.
```mT``` - sets the camera mode to timelapse video
```p.10``` - sets 10s interval
```fW``` - set FOV Wide (not Linear)
```oW1``` - Wireless connections enabled
```>r0``` - If recording > 0  
```    !19:00E``` - then End when the time hits 19:00
```!N``` - do nothing for 0.5s 
```!R11``` - repeat at position 11 (the ```>r0``` part)

**Reply by amajed172** [reply]
> If the camera where configure the only command you need is
> 
> `>r0!19:00E!N!R` That supports start at 19:01 an recording to the next 19:00
> 
> Some on your team must have done some coding. The scripting is like most computer languages, although the is simplify to fit in smaller QR Codes. The loop example simply added a print command, so you could set what was happening, and that the code is working.
> 
> Show all the cameras `*TUSB=1` once. As a separate QR Code. Nice make sure there clocks are set correctly, as they can drift. Use https://gopro.github.io/labs/control/precisiontime/
> 
> Then as you are unable to use the menus, this is an example of the full command.
> 
> `mTp.10fWoW1>r0!19:00E!N!R11`
> 
> This was the code does, on step at a time. `mT` - sets the camera mode to timelapse video `p.10` - sets 10s interval `fW` - set FOV Wide (not Linear) `oW1` - Wireless connections enabled `>r0` - If recording > 0 ` !19:00E` - then End when the time hits 19:00 `!N` - do nothing for 0.5s `!R11` - repeat at position 11 (the `>r0` part)

Thank you so much for all the time and effort you put into this. Everything works so well here
I tested it with HERO11 and HERO12 and both worked as expected

Thanks again
